### PR TITLE
Update demos

### PR DESCRIPTION
**Pull request by cremebrule**

### CODE REVIEW

**Review by ChengshuLi — COMMENTED** [review]
LGTM with some minor comments. Thanks for putting this together!

File: omnigibson/__init__.py [hunk 1]
```diff
@@ -3,6 +3,8 @@
 
 import yaml
 import builtins
+from termcolor import colored
```
[the remark this answers is not in the document]
Comment (cremebrule) [reply omnigibson/__init__.py:6]: done!

File: omnigibson/examples/object_states/particle_applier_remover_demo.py [hunk 2]
```diff
@@ -123,7 +130,7 @@ def main(random_selection=False, headless=False, short_exec=False):
     }
 
     # Sanity check inputs: Remover + Adjacency + Fluid will not work because we are using a visual_only
-    # object, so contacts will not be triggered with this object
+    # object, so contacts will not be triggered with fthis object
```
[the remark this answers is not in the document]
Comment (cremebrule) [reply omnigibson/examples/object_states/particle_applier_remover_demo.py:133]: fixed!

File: omnigibson/examples/objects/load_object_selector.py [hunk 3]
```diff
@@ -56,6 +56,11 @@ def main(random_selection=False, headless=False, short_exec=False):
     # Create the environment
     env = og.Environment(configs=cfg, action_timestep=1 / 60., physics_timestep=1 / 60.)
 
+    # Place the object so it rests on the floor
+    obj = env.scene.object_registry("name", "obj")
+    center_offset = obj.get_position() - obj.aabb_center + np.array([0, 0, obj.aabb_extent[2] / 2.0])
```
[the remark this answers is not in the document]
Comment (cremebrule) [reply omnigibson/examples/objects/load_object_selector.py:61]: Good point! The main reason i do it this way is to guarantee that the center of the object (as defined by its AABB) is always placed correctly

File: omnigibson/object_states/particle_modifier.py [hunk 4]
```diff
@@ -480,6 +487,15 @@ def physical_particle_modification_limit(self):
...
         raise NotImplementedError()
 
+    @property
+    def link(self):
+        # If self._link exists and we're using ParticleModifyMethod.ADJACENCY, we make sure it has some actual
+        # geometry to it, otherwise we use the default link
+        return self._default_link if self.method == ParticleModifyMethod.ADJACENCY and \
+                                     self._link is not None and \
+                                     len(self._link.collision_meshes) == 0 and \
+                                     len(self._link.visual_meshes) == 0 else super().link
```
[the remark this answers is not in the document]
Comment (cremebrule) [reply omnigibson/object_states/particle_modifier.py:497]: remove

File: omnigibson/simulator.py [hunk 5]
```diff
@@ -219,7 +219,7 @@ def _set_renderer_settings(self):
         carb.settings.get_settings().set_bool("/rtx/directLighting/sampledLighting/enabled", True)
         carb.settings.get_settings().set_int("/rtx/raytracing/showLights", 1)
         carb.settings.get_settings().set_float("/rtx/sceneDb/ambientLightIntensity", 0.1)
-        carb.settings.get_settings().set_int("/rtx/domeLight/upperLowerStrategy", 3)  # "Limited image-based"
+        # carb.settings.get_settings().set_int("/rtx/domeLight/upperLowerStrategy", 3)  # "Limited image-based"
```
[the remark this answers is not in the document]
Comment (cremebrule) [reply omnigibson/simulator.py:222]: done!

File: omnigibson/utils/transform_utils.py [hunk 6]
```diff
@@ -394,6 +394,19 @@ def mat2quat(rmat):
...
+    """
+    Converts given rotation matrix to quaternion.
+
+    Args:
+        rmat (np.array): (..., 3, 3) rotation matrix
+
+    Returns:
+        np.array: (..., 4) (x,y,z,w) float quaternion angles
+    """
+    return R.from_matrix(rmat).as_quat()
```
[the remark this answers is not in the document]
Comment (cremebrule) [reply omnigibson/utils/transform_utils.py:407]: done!

File: omnigibson/simulator.py [hunk 7]
```diff
@@ -613,15 +617,22 @@ def _on_contact(self, contact_headers, contact_data):
...
+                # Unique reference via uuid hashing
+                idn = actor0_obj.uuid * actor1_obj.uuid
+                if idn in combos:
+                    headers[idn].append(contact_header)
+                else:
+                    headers[idn] = []
+                    combos[idn] = (actor0_obj, actor1_obj)
+
+            for idn, headers in headers.items():
+                actor0_obj, actor1_obj = combos[idn]
```
[the remark this answers is not in the document]
Comment (cremebrule) [reply omnigibson/simulator.py:629]: timed both of them, the non-sort method is about 2x faster

[the remark this answers is not in the document]
Comment (cremebrule) [reply omnigibson/simulator.py:629]:
```python
import time
import numpy as np
a0, a1 = env.scene.objects[:2]
times_a = []
times_b = []
for i in range(1000):
    t1 = time.time()
    a0.uuid * a1.uuid
    t2 = time.time()
    tuple(sorted((a0, a1), key=lambda x: x.name))
    t3 = time.time()
    times_a.append(t2 - t1)
    times_b.append(t3 - t2)
print(f"{np.mean(times_a):.3f}, {np.mean(times_b):.3f}")
print(f"{np.sum(times_a):.3f}, {np.sum(times_b):.3f}")
```

returns:

```
0.000, 0.000
0.004, 0.009
```

File: omnigibson/systems/macro_particle_system.py [hunk 8]
```diff
@@ -746,26 +752,76 @@ def generate_group_particles_on_object(cls, group, max_samples, min_samples_for_
...
+
+        Returns:
+            2-tuple:
+                - (n, 3)-array: per-particle (x,y,z) position in the world frame
+                - (n, 4)-array: per-particle (x,y,z,w) quaternion orientation in the world frame
+        """
+        # Iterate over all particles and compute link tfs programmatically, then batch the matrix transform
+        link_tfs = dict()
+        A = np.zeros((cls.n_particles, 4, 4))
+        B = np.zeros_like(A)
```
[the remark this answers is not in the document]
Comment (cremebrule) [reply omnigibson/systems/macro_particle_system.py:770]: done!

File: omnigibson/utils/physx_utils.py [hunk 9]
```diff
@@ -238,7 +241,12 @@ def create_physx_particleset_pointinstancer(
...
     instancer_prim.GetAttribute("physxParticle:particleEnabled").Set(enabled)
 
-    if is_isosurface:
-        og.log.warning(f"Creating an instancer that uses isosurface {instancer_prim_path}. The rendering of these particles will have a delay of one timestep.")
+    # Render three more times to fully propagate changes
+    # Omni always complains about a low-level USD thing we have no control over
+    # so we suppress the warnings
+    with suppress_omni_log(channels=["omni.usd"]):
+        for i in range(3):
+            og.sim.render()
```
[the remark this answers is not in the document]
Comment (cremebrule) [reply omnigibson/utils/physx_utils.py:249]: turns out -- nope, you still need a physics step )): `og.sim.update_simulation` doesn't work either.

File: omnigibson/utils/sampling_utils.py [hunk 10]
```diff
@@ -748,11 +748,11 @@ def sample_cuboid_on_object(
                 destinations = np.array([end_pos])
 
             # Time to cast the rays.
-            cast_results = raytest_batch(start_points=sources, end_points=destinations)
+            cast_results = raytest_batch(start_points=sources, end_points=destinations, ignore_bodies=ignore_rigid_bodies)
 
             # Check whether sufficient number of rays hit the object
             hits = check_rays_hit_object(
```
[the remark this answers is not in the document]
Comment (cremebrule) [reply omnigibson/utils/sampling_utils.py:765]: done! removed args, added `verify_cuboid_empty` flag to sampling APIs

File: tests/benchmark/benchmark_interactive_scene.py [hunk 11]
```diff
@@ -42,12 +48,40 @@ def benchmark_scene(scene_name, optimized=False, import_robot=True):
...
+    # sink_filled_state = Filled(sink)
+    # sink.add_state(sink_filled_state)
+    # og.sim.play()
+    # og.sim.step()
+    # sink_filled_state.initialize(simulator=og.sim)
+    # og.sim.stop()
+    # og.sim.step()
+    # sink.states[Filled].set_value(WaterSystem, True)
+    # og.sim.step()
+    # og.sim.step()
```
Comment (ChengshuLi) [inline tests/benchmark/benchmark_interactive_scene.py:78]: are these needed?

File: tests/utils.py [hunk 12]
```diff
@@ -38,24 +35,36 @@ def get_obj_cfg(name, category, model, prim_type=PrimType.RIGID, scale=None, abi
...
+
+    # Create the scene if there's any mismatch in objects
+    create_scene = False
     if og.sim.scene is None:
+        create_scene = True
+    else:
+        for obj_cfg in objects_cfg:
+            if og.sim.scene.object_registry("name", obj_cfg["name"]) is None:
+                create_scene = True
+                break
```
Comment (ChengshuLi) [inline tests/utils.py:60]: hmm why is this needed?

Comment (cremebrule) [reply tests/utils.py:60]: pytest naively crashes because after it runs all the tests from `test_envs.py`, it then starts to try to immediately run the tests from `test_object_states.py`, without actually destroying / clearing the sim. So when `assert_test_scene()` gets called, the scene already exists so it doesn't create the new scene, and all the tests crash because none of the actual test objects exist.

File: tests/benchmark/benchmark_interactive_scene.py (hunk 11 above)
Comment (cremebrule) [reply tests/benchmark/benchmark_interactive_scene.py:78]: whoops! sorry, removed

File: tests/utils.py (hunk 12 above)
Comment (cremebrule) [reply tests/utils.py]: reverted
